fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,91 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+
+    fetch('api/todos')
+      .then(res => res.json())
+      .then(todos => {
+        this.setState({
+          todos
+        });
+      });
+  }
+
+  addTodo(newTodo) {
+
+    const newTodoStr = JSON.stringify(newTodo);
+
+    fetch('api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: newTodoStr
+    })
+      .then(res => res.json())
+      .then(data => {
+        const newTodosArr = this.state.todos;
+        newTodosArr.push(data);
+        this.setState({
+          todos: newTodosArr
+        });
+      }
+      );
+  }
+
+  toggleCompleted(todoId) {
+
+    const todoItem = this.state.todos.filter(obj => obj.id === todoId)[0];
+    const todoItemIndex = this.state.todos.indexOf(todoItem);
+    todoItem.isCompleted = !todoItem.isCompleted;
+    const todoItemStr = JSON.stringify(todoItem);
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: todoItemStr
+    })
+      .then(res => res.json())
+      .then(data => {
+        this.state.todos.splice(todoItemIndex, 1, data);
```

Comment:
In React, it's very important to not "mutate" state. If you need to change a reference data type in state, be sure to make a real copy first. Since arrays are objects (reference data), this is not a copy. However, you can use the slice method to create a copy of the array.